app/src/main/java/net/squanchy/schedule/filterschedule/FilterSchedule.kt

```diff
@@ -0,0 +1,135 @@
+
+class FilterScheduleActivity : AppCompatActivity() {
+
+    private lateinit var tracksRepository: TracksRepository
+    private lateinit var tracksFilter: TracksFilter
+    private lateinit var trackAdapter: TrackFiltersAdapter
+
+    private var subscription: Disposable? = null
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        setContentView(R.layout.activity_track_filters)
+
+        val component = tracksFilterComponent(this)
+        tracksRepository = component.tracksRepository()
+        tracksFilter = component.tracksFilter()
+
+        trackAdapter = TrackFiltersAdapter(this)
+
+        trackFiltersList.layoutManager = LinearLayoutManager(this)
+        trackFiltersList.adapter = trackAdapter
+    }
+
+    override fun onStart() {
+        super.onStart()
+
+        subscription = Observable.combineLatest(tracksRepository.tracks(), tracksFilter.selectedTracks, combineIntoCheckableTracks())
+            .subscribeOn(Schedulers.computation())
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribe {
+                trackAdapter.updateTracks(it) { track, selected ->
+                    val selectedTracks = it.filter { it.second }.map { it.first }.toSet()
+                    val newSelectedTracks = if (selected) selectedTracks + track else selectedTracks - track
+                    tracksFilter.updateSelectedTracks(newSelectedTracks)
+                    trackAdapter.notifyItemChanged(it.map { checkableTrack -> checkableTrack.first }.indexOf(track))
+                }
+            }
+    }
+
+    private fun combineIntoCheckableTracks(): BiFunction<List<Track>, Set<Track>, List<CheckableTrack>> {
+        return BiFunction { tracks, selectedTracks ->
+            tracks.map { CheckableTrack(it, selectedTracks.contains(it)) }
+        }
+    }
+
+    override fun onStop() {
+        super.onStop()
+
+        subscription?.dispose()
+    }
+}
+
+private class TrackFiltersAdapter(context: Context) : RecyclerView.Adapter<TrackViewHolder>() {
+
+    init {
+        setHasStableIds(true)
+    }
+
+    private val layoutInflater = LayoutInflater.from(context)
+
+    private var checkableTracks: List<CheckableTrack> = emptyList()
+    private var trackStateChangeListener: OnTrackSelectedChangeListener? = null
+
+    override fun getItemCount(): Int = checkableTracks.size
+
+    override fun onCreateViewHolder(parent: ViewGroup?, viewType: Int): TrackViewHolder {
+        val view = layoutInflater.inflate(R.layout.track_filters_item, parent, false) as CheckBox
+        return TrackViewHolder(view)
+    }
+
+    override fun onBindViewHolder(holder: TrackViewHolder?, position: Int) {
+        holder?.bind(checkableTracks[position], trackStateChangeListener)
+    }
+
+    override fun getItemId(position: Int): Long {
+        return checkableTracks[position].first.id.hashCode().toLong() // TODO this should use a proper checksum
```

Comment:
We should really add `numericId` to our `Track` models, but I haven't done it yet. It's trivial but would be noise, at this point. We have few elements for now so hashcode is ok.